core-plugins/src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java

```diff
@@ -0,0 +1,354 @@
+public class CreateRecordTransform extends Transform<StructuredRecord, StructuredRecord> {
+
+  /**
+   * Create hierarchy config
+   */
+  public static class CreateRecordTransformConfig extends PluginConfig {
+    public static final String FIELD_MAPPING = "fieldMapping";
+    public static final String INCLUDE_NON_MAPPED_FIELDS = "includeNonMappedFields";
+
+    @Description("Specifies the mapping for generating the hierarchy.")
+    @Name(FIELD_MAPPING)
+    String fieldMapping;
+
+    @Description("Specifies whether the fields in the input schema that are not part of the mapping, " +
+      "should be carried over as-is.")
+    @Name(INCLUDE_NON_MAPPED_FIELDS)
+    String includeNonMappedFields;
+
+    public CreateRecordTransformConfig(String fieldMapping, String includeNonMappedFields) {
+      this.fieldMapping = fieldMapping;
+      this.includeNonMappedFields = includeNonMappedFields;
+    }
+
+    public String getFieldMapping() {
+      return fieldMapping;
+    }
+
+    public JsonElement getFieldMappingJson() {
+      return GSON.fromJson(fieldMapping, JsonElement.class);
+    }
+
+    public String getIncludeNonMappedFields() {
+      return includeNonMappedFields;
+    }
+
+    public void validate(FailureCollector collector) {
+      if (!containsMacro(FIELD_MAPPING)) {
+        try {
+          getFieldMappingJson();
+        } catch (Exception e) {
+          collector.addFailure("Invalid field mapping provided.",
+                               "Please provide valid field mapping.").withConfigProperty(FIELD_MAPPING);
+        }
+      }
+      if (!containsMacro(INCLUDE_NON_MAPPED_FIELDS) && Strings.isNullOrEmpty(INCLUDE_NON_MAPPED_FIELDS)) {
+        collector.addFailure("Include missing fields property missing.",
+                             "Please provide include missing fields value.")
+          .withConfigProperty(INCLUDE_NON_MAPPED_FIELDS);
+      }
+      collector.getOrThrowException();
+    }
+  }
+
+  private final CreateRecordTransformConfig createRecordTransformConfig;
+  private static final Gson GSON = new Gson();
+  private JsonElement fieldMappingJson = null;
+
+  public CreateRecordTransform(CreateRecordTransformConfig createRecordTransformConfig) {
+    this.createRecordTransformConfig = createRecordTransformConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    createRecordTransformConfig.validate(failureCollector);
+    Schema outputSchema = null;
+    if (pipelineConfigurer.getStageConfigurer().getInputSchema() != null) {
+      //validate the input schema and get the output schema for it
+      outputSchema = getOutputSchema(pipelineConfigurer.getStageConfigurer().getInputSchema(), failureCollector);
+    }
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(outputSchema);
```

Comment:
need to handle case where fieldMapping is a macro, so output schema is not available in `configurePipeline`

Reply:
I would suggest testing with a pipeline with all configs macro enabled. Also, please make sure to handle the case where input schema is null in configurePipeline - this will happen if the previous stage's output schema is a macro.